xbmc/addons/AddonDatabase.cpp

```diff
@@ -1297,7 +1297,7 @@ bool CAddonDatabase::AddInstalledAddon(const std::shared_ptr<CAddonInfo>& addon,
 
     std::string now = CDateTime::GetCurrentDateTime().GetAsDBDateTime();
 
-    m_pDS->exec(PrepareSQL("INSERT INTO installed(addonID, enabled, installDate, origin) "
+    m_pDS->exec(PrepareSQL("INSERT OR IGNORE INTO installed(addonID, enabled, installDate, origin) "
```

Comment:
As I said I'm not an SQL guy. So, maybe a stupid question, is "OR IGNORE" standard that also works with other dbs than sqlite?

Reply:
Well, from what I read in the net, "INSERT IGNORE" is definitely not what we want as it ignores *all kind of errors" that might occur while processing the statement. 

Furthermore, "IGNORE" seems to be non-standard!

Reply:
SQL guy here, no don't do it like this.

No knowledge of the addon databse, why would insert of something that already existed be attempted? If what you really have is an add/update or add only if new situation then do what Kodi does elsewhere, check for existence in db and then update/replace or do nothing depending on what is required.

Reply:
this is non-standard (like REPLACE) and can only be seen as a workaround to get rid of the error log, because the real question is: why does `IsAddonInstalled(addonId)` return `false` -> we do run into the INSERT, while the addon is already installed (it is updated)?

tbh a question i cannot answer. `IsAddonInstalled()` is essentially a call to `GetAddon()`.

so to get this conform to standard sql we could do: 
```
SELECT * from installed where addonID = 'XYZ';
if numberofrecords = 0
  INSERT INTO.....
else
  UPDATE...
```
but that still doesn't fix the root issue and it's unlikely that we switch the Addon-database from sqlite to a different one.

@AlwinEsch what would you suggest here? perhaps i'm missing something

Reply:
Exactly your question, why does` IsAddonInstalled(addonId)` return `false` and yet the insert fails is what we need to figure out.

Until we know that we can't fix this correctly.

Reply:
`IsAddonInstalled(addonId)` returns `false` because this very moment, just before finishing the update, addonid is not yet in `m_installedAddons`. It gets removed from there during update in `CAddonMgr::UnloadAddon`and inserted again by `CAddonMgr::FindAddon`.

Reply:
1. CAddonInstallJob::DoWork() -> CAddonMgr::UnloadAddon -> m_installedAddons.erase(addonId);
2. ...
3. CAddonInstallJob::DoWork() -> CAddonMgr::LoadAddon -> CAddonMgr::FindAddon -> m_installedAddons[addonId] = it->second;

Reply:
so this means there is no other chance but renaming `AddInstalledAddon()` to `AddOrUpdateInstalledAddon()` and make the sql standard, correct? and there's no need to check if the addon is installed. let this check the db.

Reply:
CAddonMgr::IsAddonInstalled -> CAddonMgr::GetAddon -> CAddonMgr::GetAddonInfo -> auto addon = m_installedAddons.find(id);

Hope that helps.

Reply:
Debuggers and breakpoints are soooooo good if you need to find such things out. ;-)